### PR TITLE
Automatic quorum set generation

### PR DESCRIPTION
**Pull request by MonsieurNicolas**
# Description

This PR enables a new way to configure quorum sets:
the configuration now entails listing validators (with various attributes), and stellar-core computes a quorum set following best practices.
Better checks are also enabled (such as proper configuration of archives) when using this new syntax.

More details can be found in the updated documentation docs/stellar-core_example.cfg 

The sample configuration file for testnet ( docs/stellar-core_testnet.cfg ) also gives a rough idea of how configurations now look like.

### CODE REVIEW

File: src/scp/LocalNode.cpp [hunk 1]
```diff
@@ -372,17 +373,27 @@ LocalNode::findClosestVBlocking(SCPQuorumSet const& qset,
 
 Json::Value
 LocalNode::toJson(SCPQuorumSet const& qSet, bool fullKeys) const
+{
+    using std::placeholders::_1;
```
Comment (vogel) [inline src/scp/LocalNode.cpp]: does not seem to be used

File: src/main/Config.cpp [hunk 2]
```diff
@@ -286,6 +286,117 @@ Config::addHistoryArchive(std::string const& name, std::string const& get,
...
+                auto address = readString(f);
+                KNOWN_PEERS.emplace_back(address);
+            }
+            else if (f.first == "HISTORY")
+            {
+                hist = readString(f);
+            }
+            else
+            {
+                throw std::invalid_argument("malformed VALIDATORS entry");
```
Comment (vogel) [inline src/main/Config.cpp]: do we want to also expose details of why given entry is malformed here?

File: src/main/Config.cpp [hunk 3]
```diff
@@ -286,6 +286,117 @@ Config::addHistoryArchive(std::string const& name, std::string const& get,
...
+            }
+        }
+        if (ve.mName.empty() || pubKey.empty() || ve.mHomeDomain.empty())
+        {
+            throw std::invalid_argument(
+                fmt::format("malformed VALIDATORS entry {}", ve.mName));
```
Comment (vogel) [inline src/main/Config.cpp]: when condition ve.mName.empty() is true this message will not be as useful as it could be

File: src/main/Config.h [hunk 4]
```diff
@@ -28,6 +28,21 @@ struct HistoryArchiveConfiguration
 
 class Config : public std::enable_shared_from_this<Config>
 {
+    enum class ValidatorQuality : int
+    {
+        kLow = 0,
```
Comment (vogel) [inline src/main/Config.h]: Looks like you are adding 3rd convention for naming enum class entries.
Most of current ones use UPPER_CASE, one uses MixedCase (UpdateOfferResult).

Comment (MonsieurNicolas) [reply src/main/Config.h]: ah yes, good point

File: src/scp/LocalNode.cpp (hunk 1 above)
Comment (MonsieurNicolas) [reply src/scp/LocalNode.cpp]: this was deleted in a later commit, I'll clean this one up

File: src/main/Config.cpp (hunk 2 above)
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: fixed

File: src/main/Config.cpp (hunk 3 above)
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: fixed

File: src/main/Config.cpp [hunk 5]
```diff
@@ -36,6 +36,33 @@ static const std::unordered_set<std::string> TESTING_ONLY_OPTIONS = {
 static const std::unordered_set<std::string> TESTING_SUGGESTED_OPTIONS = {
     "ALLOW_LOCALHOST_FOR_TESTING"};
 
+namespace
+{
+// compute a default threshold for qset:
+// if simpleMajority is set, only require majority (>50%)
+// otherwise assume bizantine failures (~67%)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: fixed

File: src/main/Config.cpp [hunk 6]
```diff
@@ -36,6 +36,33 @@ static const std::unordered_set<std::string> TESTING_ONLY_OPTIONS = {
 static const std::unordered_set<std::string> TESTING_SUGGESTED_OPTIONS = {
     "ALLOW_LOCALHOST_FOR_TESTING"};
 
+namespace
+{
+// compute a default threshold for qset:
+// if simpleMajority is set, only require majority (>50%)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: clarified this, fixed

File: src/main/Config.cpp [hunk 7]
```diff
@@ -286,6 +286,117 @@ Config::addHistoryArchive(std::string const& name, std::string const& get,
...
+                "malformed VALIDATORS entry {} (missing quality)", ve.mName));
+        }
+        addValidatorName(pubKey, ve.mName);
+        ve.mKey = KeyUtils::fromStrKey<PublicKey>(pubKey);
+        ve.mHasHistory = !hist.empty();
+        if (ve.mHasHistory)
+        {
+            addHistoryArchive(ve.mName, hist, "", "");
+        }
+        if (ve.mQuality >= ValidatorQuality::kHigh && hist.empty())
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: I was going to add higher level, I agree it's confusing right now. fixed

File: docs/stellar-core_example.cfg [hunk 8]
```diff
@@ -363,6 +369,66 @@ MAXIMUM_LEDGER_CLOSETIME_DRIFT=50
...
+# List of validators used to automatically generate quorum sets
+# The quorum set is generated using the following rules:
+#    validators with the same home domain (representing an entity) are automatically
+#            grouped together; the threshold used assumes a simple majority (2f+1)
+#    entities are grouped by QUALITY
+#    groups are nested such that the group for the quality that precedes a given group
+#            is added as a backup for the higher quality group.
+#            ie: at top level group contains HIGH quality entities and the group that
+#              contains MEDIUM quality entities
+#    heterogeneous groups use a threshold assuming bizantine failure (3f+1)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply docs/stellar-core_example.cfg]: fixed

File: src/main/test/ConfigTests.cpp [hunk 9]
```diff
@@ -139,6 +139,61 @@ TEST_CASE("resolve node id", "[config]")
     }
 }
 
+TEST_CASE("load validators config", "[config]")
+{
+    Config c;
+    c.load("testdata/stellar-core_example_validators.cfg");
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/test/ConfigTests.cpp]: good catch

File: src/main/Config.cpp [hunk 10]
```diff
@@ -517,6 +517,28 @@ Config::load(std::string const& filename)
...
+    self.mName = "self";
+    self.mHasHistory = false;
+    if (it != domainQualityMap.end())
+    {
+        self.mQuality = it->second;
+    }
+    else
+    {
+        LOG(WARNING) << "Defaulting self to LOW priority";
+        self.mQuality = ValidatorQuality::VALIDATOR_LOW_QUALITY;
```
Comment (jonjove) [inline src/main/Config.cpp]: The choice of making this `VALIDATOR_LOW_QUALITY` seems arbitrary, so a comment is warranted here.

File: src/scp/QuorumSetUtils.cpp [hunk 11]
```diff
@@ -145,4 +147,81 @@ normalizeQSet(SCPQuorumSet& qSet, NodeID const* idToRemove)
...
+// returns -1 if l < r ; 0 if l == r ; 1 if l > r
+// lexicographical sort
+// looking at, in order: validators, innerSets, threshold
+int
+qSetCompareInt(SCPQuorumSet const& l, SCPQuorumSet const& r)
+{
+    auto& lvals = l.validators;
+    auto& rvals = r.validators;
+    auto itL = lvals.begin();
+    auto itR = rvals.end();
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/QuorumSetUtils.cpp]: no idea what happened there, fixed

File: src/scp/QuorumSetUtils.cpp [hunk 12]
```diff
@@ -106,8 +106,10 @@ isQuorumSetSane(SCPQuorumSet const& qSet, bool extraChecks)
 //  * simplifies singleton innersets
 //      { t:1, { innerSet } } into innerSet
 
+namespace
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/QuorumSetUtils.cpp]: fixed

File: src/scp/QuorumSetUtils.cpp [hunk 13]
```diff
@@ -145,4 +147,81 @@ normalizeQSet(SCPQuorumSet& qSet, NodeID const* idToRemove)
...
+    {
+        return -1;
+    }
+    if (itL != lvals.end() && itR == rvals.end())
+    {
+        return 1;
+    }
+    auto itLis = l.innerSets.begin();
+    auto itRis = r.innerSets.begin();
+    for (; itLis != l.innerSets.end() && itRis != r.innerSets.begin();
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/QuorumSetUtils.cpp]: good idea

File: src/main/Config.cpp [hunk 14]
```diff
@@ -36,6 +36,33 @@ static const std::unordered_set<std::string> TESTING_ONLY_OPTIONS = {
...
 
+namespace
+{
+// compute a default threshold for qset:
+// if simpleMajority is set and there are no inner sets, only require majority
+// (>50%) otherwise assume byzantine failures (~67%)
+unsigned int
+computeDefaultThreshold(SCPQuorumSet const& qset, bool simpleMajority)
+{
+    unsigned int res = 0;
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: yes

File: src/main/Config.cpp [hunk 15]
```diff
@@ -36,6 +36,33 @@ static const std::unordered_set<std::string> TESTING_ONLY_OPTIONS = {
...
+    unsigned int topSize =
+        (unsigned int)(qset.validators.size() + qset.innerSets.size());
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: `size_t` -> `unsigned int` requires an explicit cast

File: src/main/Config.cpp [hunk 16]
```diff
@@ -286,6 +286,125 @@ Config::addHistoryArchive(std::string const& name, std::string const& get,
...
+Config::toString(ValidatorQuality q)
+{
+    return kQualities[static_cast<int>(q)];
+}
+
+Config::ValidatorQuality
+Config::parseQuality(std::string const& q)
+{
+    auto it = std::find_if(kQualities.begin(), kQualities.end(),
+                           [&](std::string const& v) { return (v == q); });
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: fixed

File: src/main/Config.h [hunk 17]
```diff
@@ -44,6 +59,11 @@ class Config : public std::enable_shared_from_this<Config>
     void addHistoryArchive(std::string const& name, std::string const& get,
                            std::string const& put, std::string const& mkdir);
 
+    std::string toString(ValidatorQuality q);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.h]: fixed

File: src/main/Config.cpp [hunk 18]
```diff
@@ -316,7 +316,8 @@ Config::parseQuality(std::string const& q)
 
 std::vector<Config::ValidatorEntry>
 Config::parseValidators(
-    std::shared_ptr<cpptoml::base> validators)
+    std::shared_ptr<cpptoml::base> validators,
+    std::unordered_map<std::string, ValidatorQuality> domainQualityMap)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: fixed

File: src/main/Config.cpp [hunk 19]
```diff
@@ -381,6 +382,26 @@ Config::parseValidators(
             throw std::invalid_argument(
                 fmt::format("malformed VALIDATORS entry {}", ve.mName));
         }
+        auto globQualityIt = domainQualityMap.find(ve.mHomeDomain);
+        if (globQualityIt != domainQualityMap.end())
+        {
+            if (qualitySet)
+            {
+                if (ve.mQuality != globQualityIt->second)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: yeah I agree. Fixed

File: src/main/Config.cpp [hunk 20]
```diff
@@ -856,16 +857,33 @@ Config::processConfig(std::shared_ptr<cpptoml::table> t)
...
             if (qset)
             {
                 loadQset(qset->as_table(), QUORUM_SET, 0);
             }
+            auto s = toString(QUORUM_SET);
+            LOG(INFO) << "Using QUORUM_SET: " << s;
+            if (s != autoQSetStr && !validators.empty())
+            {
+                LOG(WARNING) << "Differs from generated: " << autoQSetStr;
+                if (!UNSAFE_QUORUM)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: `autoQSetStr` is not set when using the old config format, so this only happens for people that use both the new and old way at the same time (this is really used to detect when people mess up their configs somehow).

As for people that really want to mix old and new, it seems fine to just go "all manual" as this is for the "expert" users (even today, people just disable those checks most of the time as the checks are too coarse).

File: src/main/Config.cpp [hunk 21]
```diff
@@ -517,6 +517,28 @@ Config::load(std::string const& filename)
...
+    self.mHomeDomain = NODE_HOME_DOMAIN;
+    self.mName = "self";
+    self.mHasHistory = false;
+    if (it != domainQualityMap.end())
+    {
+        self.mQuality = it->second;
+    }
+    else
+    {
+        LOG(WARNING) << "Defaulting self to LOW priority";
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: ah yes, old code. fixed

[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: actually, I am making this mandatory instead

File: src/main/Config.cpp [hunk 22]
```diff
@@ -1036,7 +1070,12 @@ Config::validateConfig()
     LocalNode::forAllNodes(QUORUM_SET,
                            [&](NodeID const& n) { nodes.insert(n); });
 
-    if (nodes.size() == 0)
+    if (NODE_HOME_DOMAIN.empty() && NODE_IS_VALIDATOR)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: I removed this, checking this as part of `addSelfToValidators`

File: src/main/Config.cpp (hunk 10 above)
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: errors out now

**Comment by MonsieurNicolas** [comment]
@jonjove I addressed your comments

**Comment by MonsieurNicolas** [comment]
(and rebased, so should be ready to go)

**Comment by jonjove** [comment]
r+ 7a77a0b

**Comment by MonsieurNicolas** [comment]
forgot to update one test

**Comment by MonsieurNicolas** [comment]
r+ d781fbac461178fc139eb07a73240e490af35701

**Pull request merged**